exec/plugin-commands-script-runners/src/dlx.ts

```diff
@@ -63,44 +65,62 @@ export function help () {
 export type DlxCommandOptions = {
   package?: string[]
   shellMode?: boolean
-} & Pick<Config, 'reporter' | 'userAgent'> & add.AddCommandOptions
+} & Pick<Config, 'reporter' | 'userAgent' | 'cacheDir' | 'dlxCacheMaxAge' > & add.AddCommandOptions
 
 export async function handler (
   opts: DlxCommandOptions,
   [command, ...args]: string[]
 ) {
-  const dlxDir = await getDlxDir({
+  const pkgs = opts.package ?? [command]
+  const now = new Date()
+  const { storeDir, cachePath } = await getInfo({
     dir: opts.dir,
     pnpmHomeDir: opts.pnpmHomeDir,
     storeDir: opts.storeDir,
+    cacheDir: opts.cacheDir,
+    pkgs,
+  })
+  await fs.mkdir(cachePath, { recursive: true })
+  const { cacheLink, newPrefix, prefix } = await getPrefixInfo({
+    dlxCacheMaxAge: opts.dlxCacheMaxAge,
+    pid: process.pid,
+    cachePath,
+    now,
   })
-  const prefix = path.join(dlxDir, `dlx-${process.pid.toString()}`)
   const modulesDir = path.join(prefix, 'node_modules')
   const binsDir = path.join(modulesDir, '.bin')
-  fs.mkdirSync(prefix, { recursive: true })
-  process.on('exit', () => {
-    try {
-      fs.rmdirSync(prefix, {
-        recursive: true,
-        maxRetries: 3,
-      })
-    } catch (err) {}
-  })
-  const pkgs = opts.package ?? [command]
   const env = makeEnv({ userAgent: opts.userAgent, prependPaths: [binsDir] })
-  await add.handler({
-    // Ideally the config reader should ignore these settings when the dlx command is executed.
-    // This is a temporary solution until "@pnpm/config" is refactored.
-    ...omit(['workspaceDir', 'rootProjectManifest'], opts),
-    bin: binsDir,
-    dir: prefix,
-    lockfileDir: prefix,
-    rootProjectManifestDir: prefix, // This property won't be used as rootProjectManifest will be undefined
-    saveProd: true, // dlx will be looking for the package in the "dependencies" field!
-    saveDev: false,
-    saveOptional: false,
-    savePeer: false,
-  }, pkgs)
+  if (newPrefix) {
+    await fs.mkdir(newPrefix, { recursive: true })
+    await add.handler({
+      // Ideally the config reader should ignore these settings when the dlx command is executed.
+      // This is a temporary solution until "@pnpm/config" is refactored.
+      ...omit(['workspaceDir', 'rootProjectManifest'], opts),
+      bin: binsDir,
+      dir: newPrefix,
+      lockfileDir: newPrefix,
+      rootProjectManifestDir: newPrefix, // This property won't be used as rootProjectManifest will be undefined
+      storeDir,
+      saveProd: true, // dlx will be looking for the package in the "dependencies" field!
+      saveDev: false,
+      saveOptional: false,
+      savePeer: false,
+    }, pkgs)
+    try {
+      await fs.unlink(cacheLink)
+    } catch (err) {
+      if (!(util.types.isNativeError(err) && 'code' in err && err.code === 'ENOENT')) {
+        throw err
+      }
+    }
+    try {
+      await fs.symlink(newPrefix, cacheLink, 'junction')
```

Comment:
Use the `symlink-dir` library. It is already used in many places for creating symlinks on POSIX and junctions on Windows.

Reply:
https://github.com/pnpm/pnpm/pull/7835/commits/746a079dd3015c341c7b805f3b317dea5d226192